File: qrcode-app-php/src/Services/VisitorFingerprint.php
<?php

declare(strict_types=1);

namespace QrApp\Services;

final class VisitorFingerprint
{
    public static function compute(string $publicId, ?string $remoteIp, string $userAgent): string
    {
        $ua = $userAgent;
        if (strlen($ua) > 400) {
            $ua = substr($ua, 0, 400);
        }
        $ip = ($remoteIp === null || trim($remoteIp) === '') ? 'unknown' : trim($remoteIp);
        $raw = $publicId . '|' . $ip . '|' . $ua;

        return strtolower(hash('sha256', $raw));
    }
}

File: QRCodeApp.API/Services/VisitorFingerprint.cs
using System.Security.Cryptography;
using System.Text;

namespace QRCodeApp.API.Services
{
    /// <summary>Privacy-preserving visitor key for unique-scanner analytics (no raw PII stored).</summary>
    public static class VisitorFingerprint
    {
        public static string Compute(string publicId, string? remoteIp, string? userAgent)
        {
            var ua = userAgent ?? string.Empty;
            if (ua.Length > 400)
                ua = ua[..400];
            var ip = string.IsNullOrWhiteSpace(remoteIp) ? "unknown" : remoteIp.Trim();
            var raw = $"{publicId}|{ip}|{ua}";
            var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(raw));
            return Convert.ToHexString(bytes).ToLowerInvariant();
        }
    }
}
